File: src/main/java/com/wxl/dyttcrawler/scheduler/local/BatchPriorityScheduler.java
package com.wxl.dyttcrawler.scheduler.local;

import com.wxl.dyttcrawler.scheduler.BatchDuplicateRemovedScheduler;
import us.codecraft.webmagic.Request;
import us.codecraft.webmagic.Task;
import us.codecraft.webmagic.scheduler.MonitorableScheduler;
import us.codecraft.webmagic.utils.NumberUtils;

import java.util.Comparator;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.PriorityBlockingQueue;

/**
 * Create by wuxingle on 2020/7/20
 * 批量优先队列
 *
 * @see us.codecraft.webmagic.scheduler.PriorityScheduler
 */
public class BatchPriorityScheduler extends BatchDuplicateRemovedScheduler
        implements MonitorableScheduler {

    private static final int INITIAL_CAPACITY = 5;

    private BlockingQueue<Request> noPriorityQueue = new LinkedBlockingQueue<>();

    private PriorityBlockingQueue<Request> priorityQueuePlus = new PriorityBlockingQueue<>(
            INITIAL_CAPACITY, (o1, o2) -> -NumberUtils.compareLong(o1.getPriority(), o2.getPriority()));

    private PriorityBlockingQueue<Request> priorityQueueMinus = new PriorityBlockingQueue<>(
            INITIAL_CAPACITY, (o1, o2) -> -NumberUtils.compareLong(o1.getPriority(), o2.getPriority()));

    @Override
    public void pushWhenNoDuplicate(Request request, Task task) {
        if (request.getPriority() == 0) {
            noPriorityQueue.add(request);
        } else if (request.getPriority() > 0) {
            priorityQueuePlus.put(request);
        } else {
            priorityQueueMinus.put(request);
        }
    }

    @Override
    public synchronized Request poll(Task task) {
        Request poll = priorityQueuePlus.poll();
        if (poll != null) {
            return poll;
        }
        poll = noPriorityQueue.poll();
        if (poll != null) {
            return poll;
        }
        return priorityQueueMinus.poll();
    }

    @Override
    public int getLeftRequestsCount(Task task) {
        return noPriorityQueue.size();
    }

    @Override
    public int getTotalRequestsCount(Task task) {
        return getDuplicateRemover().getTotalRequestsCount(task);
    }
}

File: src/main/kotlin/com/wxl/dyttcrawler/scheduler/local/BatchPriorityScheduler.kt
package com.wxl.dyttcrawler.scheduler.local

import com.wxl.dyttcrawler.scheduler.BatchDuplicateRemovedScheduler
import us.codecraft.webmagic.Request
import us.codecraft.webmagic.Task
import us.codecraft.webmagic.scheduler.MonitorableScheduler
import us.codecraft.webmagic.utils.NumberUtils
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.PriorityBlockingQueue

/**
 * Create by wuxingle on 2021/10/11
 * 批量优先队列
 * @see us.codecraft.webmagic.scheduler.PriorityScheduler
 */
class BatchPriorityScheduler : BatchDuplicateRemovedScheduler(), MonitorableScheduler {

    companion object {
        const val INITIAL_CAPACITY = 5
    }

    private val noPriorityQueue = LinkedBlockingQueue<Request>()

    private val priorityQueuePlus = PriorityBlockingQueue<Request>(INITIAL_CAPACITY) { r1, r2 ->
        NumberUtils.compareLong(r1.priority, r2.priority)
    }

    private val priorityQueueMinus = PriorityBlockingQueue<Request>(INITIAL_CAPACITY) { r1, r2 ->
        NumberUtils.compareLong(r1.priority, r2.priority)
    }

    override fun pushWhenNoDuplicate(request: Request, task: Task) {
        when {
            request.priority > 0 -> priorityQueuePlus.add(request)
            request.priority < 0 -> priorityQueueMinus.add(request)
            else -> noPriorityQueue.add(request)
        }
    }

    override fun poll(task: Task): Request? {
        var req = priorityQueuePlus.poll()
        if (req != null) {
            return req
        }
        req = noPriorityQueue.poll()
        if (req != null) {
            return req
        }
        return priorityQueueMinus.poll()
    }

    override fun getLeftRequestsCount(task: Task): Int = noPriorityQueue.size

    override fun getTotalRequestsCount(task: Task): Int = duplicateRemover.getTotalRequestsCount(task)
}

